src/ai/backend/manager/models/kernel.py

```diff
@@ -154,6 +155,77 @@ def default_hostname(context) -> str:
     return f"{params['cluster_role']}{params['cluster_idx']}"
 
 
+def determine_available_next_status(status: KernelStatus) -> List[KernelStatus]:
+    blocking_statuses: List[KernelStatus]
+    match status:
+        case KernelStatus.PENDING:
+            blocking_statuses = [
+                KernelStatus.PENDING,
+            ]
+        case KernelStatus.SCHEDULED:
+            blocking_statuses = [
+                KernelStatus.PENDING,
+                KernelStatus.SCHEDULED,
+            ]
+        case KernelStatus.PREPARING:
+            blocking_statuses = [
+                KernelStatus.PENDING,
+                KernelStatus.SCHEDULED,
+                KernelStatus.PREPARING,
+            ]
+        case KernelStatus.BUILDING:
+            blocking_statuses = [
+                KernelStatus.PENDING,
+                KernelStatus.SCHEDULED,
+                KernelStatus.BUILDING,
+            ]
+        case KernelStatus.PULLING:
+            blocking_statuses = [
+                KernelStatus.PENDING,
+                KernelStatus.SCHEDULED,
+                KernelStatus.BUILDING,
+                KernelStatus.PULLING,
+            ]
+        case KernelStatus.RUNNING:
+            blocking_statuses = [
+                KernelStatus.PENDING,
+                KernelStatus.SCHEDULED,
+                KernelStatus.BUILDING,
+                KernelStatus.PULLING,
+            ]
```

Comment:
So here is the key. It ignores status update toward PREPARING if it's already RUNNING by checking this state transition table from `set_kernel_status()`. 👍🏼